### PR TITLE
Add `adapter_only` option to `save_fsdp_model` and `load_fsdp_model` to only save/load PEFT weights

### PR DESCRIPTION
**Pull request by AjayP13**
# What does this PR do?

- See related PR in `transformers` for motivation: https://github.com/huggingface/transformers/pull/28297
- Adds the ability to only save/load PEFT weights for `save_fsdp_model` and `load_fsdp_model`
- The `adapter_only` parameter (default off) controls whether only the PEFT weights will be saved.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/accelerate/blob/main/CONTRIBUTING.md#submitting-a-pull-request-pr),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/accelerate/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/accelerate/tree/main/docs#writing-documentation---specification).
- [ ] Did you write any new necessary tests?

### CODE REVIEW

File: src/accelerate/utils/fsdp_utils.py [hunk 1]
```diff
@@ -32,7 +32,32 @@
 logger = get_logger(__name__)
 
 
-def save_fsdp_model(fsdp_plugin, accelerator, model, output_dir, model_index=0):
+def _is_peft_model(model):
+    if is_peft_available():
+        from peft import PeftModel
+    unwrapped_model = getattr(model.module, "_orig_mod", model.module)
```
Comment (younesbelkada) [inline src/accelerate/utils/fsdp_utils.py:38]: do we have an unwrap model utility method in accelerate? If `model` is not a DDP `model.module` will fail here no?

**Review by younesbelkada — COMMENTED** [review]
Thanks a lot ! I left one comment - I will let @pacman100 give his final opinion on the PR as he is more familiar with FSDP and accelerate

File: src/accelerate/utils/fsdp_utils.py (hunk 1 above)
Comment (younesbelkada) [reply src/accelerate/utils/fsdp_utils.py:38]: ok perfect then! maybe worht mentioning in a comment that `_is_peft_model` is only meant to be used for FSDP models, or maybe change the method name to `_is_fsdp_peft_model`

Comment (muellerzr) [reply src/accelerate/utils/fsdp_utils.py:38]: I think `extract_model_from_parallel` in accelerate should cover this case if desired

Comment (muellerzr) [reply src/accelerate/utils/fsdp_utils.py]: But worth double checking/writing a test :)

Comment (AjayP13) [reply src/accelerate/utils/fsdp_utils.py]: Thanks @muellerzr, I've just updated this with the unwrapping utility and re-ran the test we have on the other branch on the multiple-GPU machine and it works.

**Review by pacman100 — COMMENTED** [review]
Thank you @AjayP13, the changes in this PR make sense to me. However, they have been only tested with the `FULL_STATE_DICT` state dict type, would this work with `SHARDED_STATE_DICT` too? If not, then this logic should only be limited to the cases wherein `FULL_STATE_DICT` is used.

**Comment by AjayP13** [comment]
> Thank you @AjayP13, the changes in this PR make sense to me. However, they have been only tested with the `FULL_STATE_DICT` state dict type, would this work with `SHARDED_STATE_DICT` too? If not, then this logic should only be limited to the cases wherein `FULL_STATE_DICT` is used.

Thanks for the review @pacman100 , I just tested this, and this works with `SHARDED_STATE_DICT` as well (similarly the loss drops, sharded state dicts per worker get saved to disk, resuming & load_best both work).

**Review by pacman100 — APPROVED** [review]
Thank you @AjayP13 for enabling only the storage of adapter weights when using PEFT+FSDP, very useful! ✨

File: src/accelerate/utils/fsdp_utils.py [hunk 2]
```diff
@@ -32,7 +33,31 @@
...
+        return model.state_dict()
+
+
+def _set_model_state_dict(model, state_dict, adapter_only=False):
+    if adapter_only and _is_peft_model(model):
+        from peft import set_peft_model_state_dict
+
+        return set_peft_model_state_dict(model, state_dict, adapter_name=model.active_adapter)
+    else:
+        return model.load_state_dict(state_dict)
```
Comment (muellerzr) [inline src/accelerate/utils/fsdp_utils.py:57]: I'm not the biggest fan of this pattern we're working with here. 

These should live inside of one function, and currently from what I can tell we have 2 different ways of getting the state_dict inside accelerate now:
1. Here
2. `Accelerator.get_state_dict`

This hints at me that perhaps we need to chunk up `Accelerator`'s `get_state_dict` into something that can be called elsewhere (and stored in `utils.modeling` probably). 

Same with `_set_state_dict` as well. While yes it's just FSDP, then we can write a check for that if necessary. 

Secondly: we generally don't follow the practice of hidden function names in Accelerate, and everything should be made public unless for *extreme* circumstances, and I'm not convinced this is one of those. 

Can we rewrite this a bit to make it more extensible? 

Perhaps just do an `if/else` for calling `set_peft_model_state_dict` or not, and just importing them at the top of the file. This way we can avoid this entirely.

File: src/accelerate/utils/fsdp_utils.py [hunk 3]
```diff
@@ -32,7 +33,31 @@
 logger = get_logger(__name__)
 
 
-def save_fsdp_model(fsdp_plugin, accelerator, model, output_dir, model_index=0):
+def _is_peft_model(model):
+    if is_peft_available():
+        from peft import PeftModel
+    return is_peft_available() and isinstance(extract_model_from_parallel(model), PeftModel)
```
Comment (muellerzr) [inline src/accelerate/utils/fsdp_utils.py:39]: Let's move this entire function to `utils.modeling` please, as that makes the most logical sense for it

File: src/accelerate/utils/imports.py [hunk 4]
```diff
@@ -175,6 +175,10 @@ def is_datasets_available():
     return _is_package_available("datasets")
 
 
+def is_peft_available():
```
Comment (muellerzr) [inline src/accelerate/utils/imports.py:178]: We need to add this to `utils.__init__.py`

**Review by muellerzr — COMMENTED** [review]
Thanks! Overall this seems very helpful, but let's work on the design a bit to adhear to more of our practices. Left some comments and suggestions 😄

File: src/accelerate/utils/imports.py (hunk 4 above)
Comment (AjayP13) [reply src/accelerate/utils/imports.py:178]: Done.

File: src/accelerate/utils/fsdp_utils.py (hunk 3 above)
Comment (AjayP13) [reply src/accelerate/utils/fsdp_utils.py]: Done.

File: src/accelerate/utils/fsdp_utils.py (hunk 2 above)
Comment (AjayP13) [reply src/accelerate/utils/fsdp_utils.py]: @muellerzr See the newest refactor, but I believe this is not possible.

`Accelerator.get_state_dict` already has logic for FSDP, but it handles a different use case for getting the FSDP state dict. `Accelerator.get_state_dict` always returns the `FULL_STATE_DICT` from FSDP. Meanwhile, this file, gets the state dict according to how the user configured they want it to be *saved* (could be a `SHARDED_STATE_DICT`). I believe that is why this file originally never called `Accelerator.get_state_dict`.

As for your suggestion of importing `get_peft_model_state_dict` at the top v.s. encapsulated in the hidden function, we can't import at the top because `accelerate` does not have a dependency on `peft` and importing at the top would throw an error if `peft` is not available on a users computer. Also, the hidden function `_get_model_state_dict` is used 3 times in this file. I could get rid of the hidden function and replace with an inline if-statement in those 3 places, but it would introduce 3 places of repeated code that would need to be updated together. Between the repeated code and the need for keeping the `peft` imports local v.s. top-of-file imports, I think it may make sense to keep the two hidden functions here.

Comment (AjayP13) [reply src/accelerate/utils/fsdp_utils.py]: @muellerzr Have you had a chance to look this over? Waiting on this PR to get https://github.com/huggingface/transformers/pull/28297 merged.

**Review by muellerzr — APPROVED** [review]
Thanks, your comments make sense!
